nbextensions/appCell2/widgets/tabs/resultsViewer.js

```diff
@@ -15,119 +15,170 @@ define([
-                    const tag = model.getItem('app.tag');
-                    if (suggestions.next_methods) {
-                        return nms.get_method_spec({
-                            ids: suggestions.next_methods,
-                            tag,
-                        });
-                    }
-                })
-                .then((nextApps) => {
-                    renderNextApps(nextApps);
+            // this way first - if this is a parent job of a batch, and we have
+            // a report_name in the result, then show the batch result
+            if (isParentJob && result && result.params && result.params.report_name) {
+                reportParams = result.params;
+            } else if (
+                jobState.widget_info &&
+                jobState.widget_info.params &&
+                jobState.widget_info.params.report_name
+            ) {
+                // otherwise, if there's some report info in the jobState, show that
+                // report
+                reportParams = jobState.widget_info.params;
+            }
+
+            // if there are no report parameters, then just dump the info given in
+            // the job output
+            if (!reportParams) {
+                return Promise.try(() => {
+                    const jobOutput = jobState.job_output
+                        ? jobState.job_output.result
+                        : 'no output found';
+                    ui.getElement('results').classList.remove('hidden');
+                    ui.setContent('results.body', ui.buildPresentableJson(jobOutput));
                 });
+            }
+            // otherwise, render the report;lp
+            return renderReportView(reportParams);
         }
 
-        function lazyRenderReport() {
-            const nbContainer = document.querySelector('#notebook-container');
-            // Add scroll event listener to the notebook container on first call.
-            if (!reportRenderTimeout) {
-                nbContainer.addEventListener('scroll', lazyRenderReport);
-            }
-            // Use a debounce timeout to avoid rendering the report _while_ the user is scrolling.
-            clearTimeout(reportRenderTimeout);
-            reportRenderTimeout = setTimeout(() => {
-                const reportElem = ui.getElement('report-widget');
-                // Once scrolling stops...
-                if (!reportRendered && DisplayUtil.verticalInViewport(reportElem)) {
-                    new KBaseReportView($(reportElem), reportParams);
-                    reportRendered = true;
-                }
-                if (reportRendered) {
-                    // Remove the scroll event listener if the report has been rendered.
-                    nbContainer.removeEventListener('scroll', lazyRenderReport);
+        function lazyRenderReport(reportParams) {
+            return Promise.try(() => {
+                const nbContainer = document.querySelector('#notebook-container');
```

Comment:
Also mostly unchanged, just wrapped in a Promise. I'm not sure this is the best way, but it keeps the chains working right.